---
title: "README"
author: ""
date: "04/13/2024"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Experiments with simulated $W$ and $\pi_i$’s
```{r sim1}
# please specify the path
path="~/Dropbox/research/decals/DECALS/"
source(paste0(path,"code/DECALS.R"))
load(file=paste0(path,"data/sim1/cp_decals.RData"))
load(file=paste0(path,"data/sim1/cp_ols.RData"))
load(file=paste0(path,"data/sim1/cp_mead.RData"))
# These data are generated by 'simulation1.R'.
```



```{r sim1plot, echo=FALSE}
library(vioplot)
par(mfrow=c(1,3))
for(k in 1:3){
  result=cbind(cp_ols[,k],cp_mead[,k],cp_decals[,k])
  colnames(result)=c("OLS","MEAD","DECALS")
  vioplot(result,col=c("grey","lightcoral","skyblue"),#axes=FALSE,
          main=paste("Cell",k),ylim=c(0.6,1),outline=FALSE,
          ylab="coverage probability")
  lines(c(0.5,3.5),c(0.95,0.95),lty=2)
}
par(mfrow=c(1,1))
rm(cp_decals);rm(cp_mead);rm(cp_ols)
```

```{r sensitive, echo=FALSE}
# Sensitivity analysis
cell1=cell2=cell3=NULL
for(j in 1:10){
  load(file=paste0(path,"data/sim1/noise",j,"/cp_decals.RData"))
  cell1=cbind(cell1,cp_decals[,1])
  cell2=cbind(cell2,cp_decals[,2])
  cell3=cbind(cell3,cp_decals[,3])
}
colnames(cell1)=colnames(cell2)=colnames(cell3)=seq(0.1,1,by=0.1)
noise_cp=list(cell1=cell1,cell2=cell2,cell3=cell3)

for(k in 1:3){
  result=noise_cp[[k]]
  vioplot(result,col=c("skyblue"),#axes=FALSE,
          main=paste("Cell",k),ylim=c(0,1),outline=FALSE,
          ylab="coverage probability")
  lines(c(0.5,10.5),c(0.95,0.95),lty=2)
}

```


```{r gls, echo=FALSE}
# the figure for generalized least square in Section 2.4
load(file=paste0(path,"data/sim1/gls_cp.RData"))
par(mfrow=c(1,3))
for(k in 1:3){
  result=gls_cp[[k]]
  colnames(result)=c("OLS true","GLS true","OLS est","GLS est")
  vioplot(result,col=c("mistyrose","skyblue"),axes=FALSE,
          main=paste("Cell",k),ylim=c(0.3,1),#outline=FALSE,
          ylab="coverage probability")
  lines(c(0.5,4.5),c(0.95,0.95),lty=2)
}
par(mfrow=c(1,1))
```



## Experiments with $W$ and $\pi_i$’s inferred from real data
```{r sim2}
load(file=paste0(path,"data/sim2/cp_decals.RData"))
load(file=paste0(path,"data/sim2/cp_ols.RData"))
load(file=paste0(path,"data/sim2/cp_mead.RData"))
load(file=paste0(path,"data/sim2/cp_rna_sieve.RData"))
# These data are generated by 'simulation2.R'.
```


```{r sim2plot, echo=FALSE}
par(mfrow=c(2,3))
for(k in 1:5){
  result=cbind(cp_ols[,k],cp_rna_sieve[,k],cp_mead[,k],cp_decals[,k])
  colnames(result)=c("OLS","RNA-Sieve","MEAD","DECALS")
  vioplot(result,col=c("grey","lightgreen","lightcoral","skyblue"),#axes=FALSE,
          main=paste("Cell",k),ylim=c(0,1),outline=FALSE,
          ylab="coverage probability")
  lines(c(0.5,3.5),c(0.95,0.95),lty=2)
}
par(mfrow=c(1,1))
```


```{r sim2gaussian}
load(file=paste0(path,"data/sim2/gaussian/cp_decals.RData"))
load(file=paste0(path,"data/sim2/gaussian/cp_ols.RData"))
load(file=paste0(path,"data/sim2/gaussian/cp_mead.RData"))
load(file=paste0(path,"data/sim2/gaussian/cp_rna_sieve.RData"))
# These data are generated by the Gaussian part of 'simulation2.R'.
```


```{r sim2gaussianplot, echo=FALSE}
par(mfrow=c(2,3))
for(k in 1:5){
  result=cbind(cp_ols[,k],cp_rna_sieve[,k],cp_mead[,k],cp_decals[,k])
  colnames(result)=c("OLS","RNA-Sieve","MEAD","DECALS")
  vioplot(result,col=c("grey","lightgreen","lightcoral","skyblue"),#axes=FALSE,
          main=paste("Cell",k),ylim=c(0,1),outline=FALSE,
          ylab="coverage probability")
  lines(c(0.5,3.5),c(0.95,0.95),lty=2)
}
par(mfrow=c(1,1))
```

## ROSMAP Data
```{r rosmap}
load(file=paste0(path,"data/rosmap/bulk.RData"))
load(file=paste0(path,"data/rosmap/sig.RData"))
CTS_proportion=constraint_ols(sig=sig,bulk=bulk)
dim(CTS_proportion) # samples * CTS_proportions
# These data are generated by 'rosmap.R'.
# The result for bMIND analysis is given in "data/rosmap".
# The analysis for DE genes are given in 'rosmap.R'.
```


```{r rosmapplot, echo=FALSE}
cell_names=c("Neu","Oli","Ast","Mic","End")
colnames(CTS_proportion)=cell_names
boxplot(CTS_proportion,outline=FALSE,ylab="cell type proportions",ylim=c(0,0.8))
```

## GTEx Data
```{r gtex}
load(file=paste0(path,"data/gtex/bulk.RData"))
load(file=paste0(path,"data/gtex/sig.RData"))
CTS_proportion=constraint_ols(sig=sig,bulk=bulk)
dim(CTS_proportion) # samples * CTS_proportions
# These data are generated by 'gtex.R'.
# The result for bMIND analysis is given in "data/gtex".
# The analysis for DE genes are given in 'gtex.R'.
```


```{r gtexplot, echo=FALSE}
cell_names=c("Ast","End","Mic","Ext","Inh","Pli")
colnames(CTS_proportion)=cell_names
boxplot(CTS_proportion,outline=FALSE,ylab="cell type proportions",ylim=c(0,0.8))
```


```{r gtexbar}
load(file=paste0(path,"data/gtex/x_genes.RData"))
load(file=paste0(path,"data/gtex/y_genes.RData"))
load(file=paste0(path,"data/gtex/CTS_chr_sd.RData"))
load(file=paste0(path,"data/gtex/x_sex.RData"))
load(file=paste0(path,"data/gtex/y_sex.RData"))
load(file=paste0(path,"data/gtex/sex_de_sd.RData"))
# These data are generated by 'rosmap.R'.
```


```{r gtexbarplot, echo=FALSE}
par(mfrow=c(2,2))
barplot(x_genes, beside=TRUE,
        main = "Proportions of DE transcripts on Chromosome X",
        #xlab = "Class",
        col = c("lightgreen","lightblue"),ylim=c(0,0.18)
)
for(k in 1:6){
  lines(x=c((3*k-1)+0.3,(3*k-1)+0.7),y=rep(x_genes[2,k]-2*CTS_chr_sd[k,1],2),lty=1,col="red")
  lines(x=c((3*k-1)+0.3,(3*k-1)+0.7),y=rep(x_genes[2,k]+2*CTS_chr_sd[k,1],2),lty=1,col="red")
  lines(x=c((3*k-1)+0.5,(3*k-1)+0.5),y=c(x_genes[2,k]-2*CTS_chr_sd[k,1],x_genes[2,k]+2*CTS_chr_sd[k,1]),lty=2,col="red")
}

barplot(y_genes, beside=TRUE,
        main = "Ratio of DE genes in Chromosome Y",
        #xlab = "Class",
        col = c("lightgreen","lightblue"),ylim=c(0,0.55)
)
for(k in 1:6){
  lines(x=c((3*k-1)+0.3,(3*k-1)+0.7),y=rep(y_genes[2,k]-2*CTS_chr_sd[k,2],2),lty=1,col="red")
  lines(x=c((3*k-1)+0.3,(3*k-1)+0.7),y=rep(y_genes[2,k]+2*CTS_chr_sd[k,2],2),lty=1,col="red")
  lines(x=c((3*k-1)+0.5,(3*k-1)+0.5),y=c(y_genes[2,k]-2*CTS_chr_sd[k,2],y_genes[2,k]+2*CTS_chr_sd[k,2]),lty=2,col="red")
}


barplot(x_sex, beside=TRUE,
        #main = "Chromosome X: the ratio of overexpressed DE genes in female",
        #xlab = "Class",
        col = c("lightgreen","lightblue"),ylim=c(0,1.02)
)
for(k in 1:6){
  lines(x=c((3*k-1)+0.3,(3*k-1)+0.7),y=rep(x_sex[2,k]-2*sex_de_sd[k,1],2),lty=1,col="red")
  lines(x=c((3*k-1)+0.3,(3*k-1)+0.7),y=rep(x_sex[2,k]+2*sex_de_sd[k,1],2),lty=1,col="red")
  lines(x=c((3*k-1)+0.5,(3*k-1)+0.5),y=c(x_sex[2,k]-2*sex_de_sd[k,1],x_sex[2,k]+2*sex_de_sd[k,1]),lty=2,col="red")
}
barplot(y_sex, beside=TRUE,
        #main = "Chromosome Y: the ratio of overexpressed DE genes in male",
        col = c("lightgreen","lightblue")
)
for(k in 1:6){
  lines(x=c((3*k-1)+0.3,(3*k-1)+0.7),y=rep(y_sex[2,k]-2*sex_de_sd[k,2],2),lty=1,col="red")
  lines(x=c((3*k-1)+0.3,(3*k-1)+0.7),y=rep(y_sex[2,k]+2*sex_de_sd[k,2],2),lty=1,col="red")
  lines(x=c((3*k-1)+0.5,(3*k-1)+0.5),y=c(y_sex[2,k]-2*sex_de_sd[k,2],y_sex[2,k]+2*sex_de_sd[k,2]),lty=2,col="red")
}
par(mfrow=c(1,1))
```

## All of these can be ontained \url{https://github.com/BiaoCai1993/DECALS/tree/main}.